src/midi/fluid_midi.c

```diff
@@ -2884,3 +3005,414 @@ fluid_midi_event_length(unsigned char event)
+			return FLUID_FAILED;
+		}
+		player->transposed_ch[chan] = FALSE;
+		FLUID_LOG(FLUID_DBG, "Disabled transposition on channel %d\n", chan);
+		return FLUID_OK;
+}
+
+/* Returns 1 if transposed, 0 if not */
+int fluid_player_get_channel_transpose(fluid_player_t *player, unsigned char chan)
+{
+	return player->transposed_ch[chan];
+}
+
+/* Returns 1 if transposed, 0 if not */
+int fluid_player_get_transpose(fluid_player_t *player)
+{
+	return player->transpose;
+}
+
+/* Returns 1 if muted, 0 if not -1 if channel is invalid */
+int fluid_player_get_channel_mute(fluid_player_t *player, unsigned char chan)
+{
+	if(chan > MAX_NUMBER_OF_CHANNELS) {return -1;}
+	return player->muted_ch[chan];
+}
+
+/* Returns 1 if muted, 0 if not, -1 if track is invalid */
+int fluid_player_get_track_mute(fluid_player_t *player, unsigned char track)
+{
+	if(track > player->ntracks) {return -1;}
+	return player->muted_tracks[track];
+}
+
+/**
+ * Get the program number on a MIDI channel.
+ * @param player FluidSynth player instance
+ * @param chan MIDI channel number (0 to MIDI channel count - 1)
+ * @return program number (0-127) 
+   #FLUID_FAILED if channel is invalid, or in other fail cases.
+ */
+int fluid_player_get_program(fluid_player_t *player, unsigned char chan)
+{
+	int program;
+	if(chan > MAX_NUMBER_OF_CHANNELS) {return FLUID_FAILED;}
+	fluid_synth_get_program(player->synth, chan, NULL, NULL, &program);
+	return program;
+}
+
+/**
+ * Get the Control Change value on a MIDI channel.
+ * @param player FluidSynth player instance
+ * @param chan MIDI channel number (0 to MIDI channel count - 1)
+ * @return MIDI Control Change value (0-127) 
+   #FLUID_FAILED if channel is invalid
+ */
+int fluid_player_get_cc(fluid_player_t *player, unsigned char chan, unsigned char cc_num)
+{
+	int pval;
+	if(chan > MAX_NUMBER_OF_CHANNELS) {return FLUID_FAILED;}
+	fluid_synth_get_cc(player->synth, chan, cc_num, &pval);
+	return pval;
+}
+
+/**
+ * Get the MIDI pitch bend controller value on a MIDI channel.
+ * @param player FluidSynth player instance
+ * @param chan MIDI channel number (0 to MIDI channel count - 1)
+ * @return MIDI pitch bend value (0-16383 with 8192 being center), 
+   #FLUID_FAILED if channel is invalid
+ */
+int fluid_player_get_pitch_bend(fluid_player_t *player, unsigned char chan)
+{
+	int pbend;
+	if(chan > MAX_NUMBER_OF_CHANNELS) {return FLUID_FAILED;}
+	fluid_synth_get_pitch_bend(player->synth, chan, &pbend);
+	return pbend;
+}
+
+/*Loop handler for callback. Always returns FLUID_OK. */
+
+int fluid_player_handle_loops(void *data, int tick)
+{
+	fluid_player_t *player = (fluid_player_t *)data;
+
+	int total_ticks = fluid_player_get_total_ticks(player) - 1;	/* subtract one for safety */
+	
+	/* return if loopstart tick is undefined in MIDI or */
+	/* loopend tick is undefined in MIDI and player hasn't reached song end tick */
+	if(!player->loopstart_tick || (!player->loopend_tick && tick < total_ticks))
```

Comment:
So this means that if your loopstart is at the very beginning, the very first tick of the MIDI file, your looping logic won't work, wouldn't it?